Service Fabric 9.0 Release Notes (#1346)
microsoft/service-fabric: release_notes/Service_Fabric_ReleaseNotes_90.md

# Microsoft Azure Service Fabric 9.0 Release Notes

> This release includes the bug fixes and features described in this document. This release includes runtime, SDKs, and Windows Server Standalone deployments to run on-premises.

Reviewer: ```suggestion
This release includes the bug fixes and features described in this document. This release also includes runtime, SDKs, and Windows Server Standalone deployments to run on-premises.
```

The following packages and versions are part of this release:

| Service | Platform | Version |
|---------|----------|---------|
|Service Fabric Runtime| Ubuntu 18.04 <br> Ubuntu 20.04 <br> Windows | 9.0.1018.1 <br> 9.0.1018.1 <br> 9.0.1017.9590 |
|Service Fabric for Windows Server|Service Fabric Standalone Installer Package | 9.0.1017.9590 |
|.NET SDK |Windows .NET SDK <br> Microsoft.ServiceFabric <br> Reliable Services and Reliable Actors <br> ASP.NET Core Service Fabric integration| 6.0.1017 <br> 9.0.1017 <br> 9.0.1017 <br> 9.0.1017 |
|Java SDK |Java for Linux SDK | 1.0.6 |
|Service Fabric PowerShell and CLI | AzureRM PowerShell Module <br> SFCTL | 0.3.15 <br> 11.1.0 |

<br>

## Contents 

Microsoft Azure Service Fabric 9.0 Release Notes

* [Key Announcements](#key-announcements)
* [Service Fabric Runtime](#service-fabric-runtime)
* [Service Fabric Common Bug Fixes](#service-fabric-common-bug-fixes)
* [Future Retirement and Deprecation Path Callouts](#future-retirement-and-deprecation-path-callouts)
* [Current And Upcoming Breaking Changes](#Current-And-Upcoming-Breaking-Changes)
* [Repositories and Download Links](#repositories-and-download-links)

<br>

## Key Announcements

> * [**General Availability**] Support for .NET 6.0

Reviewer 2: Does this include any updates to the SDKs to support .NET 6.0?

* [**General Availability**] Support for Ubuntu 20.04

> * [**General Availability**] Support for Multi-AZ VMSS

Reviewer 2: Can you specify that this is multi az spanning in a single VMSS?

Read more about these key announcements in the sections listed below.

<br>

## Service Fabric Runtime

All features and fixes listed in the table below are targeted for the following package versions:<br>

| OS | Service Fabric Runtime Version |
|----|--------------------------------|
| **Windows** | **9.0.1017.9590** |
| **Ubuntu 18.04** | **9.0.1018.1** |
| **Ubuntu 20.04** | **9.0.1018.1** |

<br>

| S# | IssueType | Description | Resolution | 
|----------|-----------|-------------|------------|

> | 1. | **Feature** | [**General Availability**] Support for .NET 6.0 | **Brief Description**: Support for .NET 6.0 has been in preview since Service Fabric runtime v8.2 and Service Fabric SDK v5.2. With Service Fabric runtime v9.0 and Service Fabric SDK v.6.0 release, this has now been made generally available.<br>Please read futher feature section in this table below for features that ship with this support, including IHost. IHostBuilder support, Data Contract Serializer support, etc.<br>Support for .NET 6.0 for Service Fabric has been made available in Visual Studio 2022 v7.1 onwards. To enable using this feature is [Visual Studio 2022 version 17.1](https://docs.microsoft.com/en-us/visualstudio/releases/2022/release-notes) or beyond, just install the Service Fabric SDK v6.0 or greater.<br>**Impact**: Service Fabric application running on .NET 5.0 can now be upgraded to .NET 6.0.<br>**Documentation Reference**: Read more about it at -<ul><li>[Service Fabric supported versions](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-versions)</li><li>Also, see feature *Added support for IHost, IHostBuilder and Minimal Hosting Model* listed below</li><li>And, see feature *Enabling opt-in option for Data Contract Serialization (DCS) based remoting exception* listed below</li></ul> |

Reviewer: ^Just some typo fixes and rephrasing

> | 3. | **Feature** | [**General Availability**] Support for Multi-AZ VMSS | **Brief Description**: AZ Spanning Virtual Machine Scale Sets (VMSS) has been in preview for a year becasue we did not have the ability to orchestrate VMSS updates, one AZ at a time until now.<br>We can have the ability to drive VMSS updates, sequentially 1 zone at a time by setting VMSSZonalUpgradeMode = "Hierarchical". This value can be changed at anytime.<br>**Documentation Reference**: Read more about it at -<ul><li>[Deploy an Azure Service Fabric cluster across Availability Zones](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cross-availability-zones)</li></ul> |

Reviewer 2: AZ Spanning Virtual Machine Scale Sets (VMSS) has been in preview for a year becasue we did not have the ability to orchestrate VMSS updates, one AZ at a time until now.  

I would recommend phrasing as: 
AZ Spanning Virtual Machine Scale Sets (VMSS) has been in preview and is now entering general availability. 

We can have the ability to drive VMSS updates, sequentially 1 zone at a time by setting VMSSZonalUpgradeMode = "Hierarchical". This value can be changed at anytime.

-> 
VMSS upgrades can now be configured to occur sequentially 1 zone at a time by setting VMSSZonalUpgradeMode = "Hierarchical". This value can be updated on the VMSS at anytime after creation.

> | 4. | **Feature** | Added support for IHost, IHostBuilder and Minimal Hosting Model | **Brief Description**: IHost integration in KestrelCommunicationListener and HttpSysCommunicationListener was shipped in SDK v5.2.1363 in Preview. In Service Fabric 9.0, this is made generally available.<br>Add IHost and Minimal Hosting based web service samples in the documentation.<br>In addition to IWebHost/IWebHostBuilder, KestrelCommunicationListener and HttpSysCommunicationListener now support building ASP.NET Core web services using IHost/IHostBuilder.<br>.Net 6 introduced the Minimal Hosting model which is a more simplified and streamlined way of creating ASP.NET Core web applications. Minimal hosting model can also be used for creating SF ASP.NET Core services.<br>This was available in preview starting with v5.2.1363 of Microsoft.ServiceFabric.AspNetCore.Kestrel and Microsoft.ServiceFabric.AspNetCore.HttpSys packages. Starting Service Fabric v.9.0, this has been made generally available.<br>**Documentation Reference**: Read more about it at -<ul><li>[ASP.NET Core in Azure Service Fabric Reliable Services](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore)</li><li>[IHost and Minimal Hosting integration](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore#ihost-and-minimal-hosting-integration)</li></ul> |

Reviewer 2: Add IHost and Minimal Hosting based web service samples in the documentation -> Is this supposed to be replaced by a sample or documentation link?

> | 4. | **Feature** | Added support for IHost, IHostBuilder and Minimal Hosting Model | **Brief Description**: IHost integration in KestrelCommunicationListener and HttpSysCommunicationListener was shipped in SDK v5.2.1363 in Preview. In Service Fabric 9.0, this is made generally available.<br>Add IHost and Minimal Hosting based web service samples in the documentation.<br>In addition to IWebHost/IWebHostBuilder, KestrelCommunicationListener and HttpSysCommunicationListener now support building ASP.NET Core web services using IHost/IHostBuilder.<br>.Net 6 introduced the Minimal Hosting model which is a more simplified and streamlined way of creating ASP.NET Core web applications. Minimal hosting model can also be used for creating SF ASP.NET Core services.<br>This was available in preview starting with v5.2.1363 of Microsoft.ServiceFabric.AspNetCore.Kestrel and Microsoft.ServiceFabric.AspNetCore.HttpSys packages. Starting Service Fabric v.9.0, this has been made generally available.<br>**Documentation Reference**: Read more about it at -<ul><li>[ASP.NET Core in Azure Service Fabric Reliable Services](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore)</li><li>[IHost and Minimal Hosting integration](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore#ihost-and-minimal-hosting-integration)</li></ul> |

Reviewer: ```suggestion
| 4. | **Feature** | Added support for IHost, IHostBuilder and Minimal Hosting Model | **Brief Description**: IHost integration in KestrelCommunicationListener and HttpSysCommunicationListener was shipped in SDK v5.2.1363 in Preview. In Service Fabric 9.0, this is made generally available.<br>Add IHost and Minimal Hosting based web service samples in the documentation.<br>In addition to IWebHost/IWebHostBuilder, KestrelCommunicationListener and HttpSysCommunicationListener now support building ASP.NET Core web services using IHost/IHostBuilder.<br>.Net 6 introduced the Minimal Hosting model which is a more simplified and streamlined way of creating ASP.NET Core web applications. Minimal hosting model can also be used for creating SF ASP.NET Core services.<br>This was available in preview starting with v5.2.1363 of Microsoft.ServiceFabric.AspNetCore.Kestrel and Microsoft.ServiceFabric.AspNetCore.HttpSys packages. Starting with Service Fabric v.9.0, this has been made generally available.<br>**Documentation Reference**: Read more about it at -<ul><li>[ASP.NET Core in Azure Service Fabric Reliable Services](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore)</li><li>[IHost and Minimal Hosting integration](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore#ihost-and-minimal-hosting-integration)</li></ul> |
```

Reviewer: ^"Starting Service Fabric v.9.0" --> "Starting with Service Fabric v.9.0"

> | 5. | **Feature** | Enabling opt-in option for Data Contract Serialization (DCS) based remoting exception | **Brief Description**: BinaryFormatter is the default exception serializer for v2 remoting. AspNetCore 5 blocks BinaryFormatter serialization by default.<br>Starting Aspnet 5.0, BinaryFormatter based serialization is blocked by default. This change affected exception messaging in remoting services. If the service code is dependent on remoting exception for exception handling, retry logic, etc., then user would need to use EnableUnsafeBinaryFormatterSerialization in their aspnet project file.<br>**Impact**: This feature introduces an opt in method to enable DataContract remoting exception serialization.<br>**Solution**: Remoting V2 introduces DataContract remoting exception serialization. User could opt this feature by using abricTransportRemotingListenerSettings.ExceptionSerializationTechnique on the Service side and FabricTransportRemotingSettings.ExceptionDeserializationTechnique on the client side.<br>**Documentation Reference**: Read more about it at -<ul><li>[Enabling Data Contract serialization for Remoting exceptions in Service Fabric](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-exception-serialization)</li></ul> |

Reviewer: ```suggestion
| 5. | **Feature** | Enabling opt-in option for Data Contract Serialization (DCS) based remoting exception | **Brief Description**: BinaryFormatter is the default exception serializer for v2 remoting. AspNetCore 5 blocks BinaryFormatter serialization by default.<br>Starting with Aspnet 5.0, BinaryFormatter based serialization is blocked by default. This change affected exception messaging in remoting services. If the service code is dependent on remoting exception for exception handling, retry logic, etc., then the user would need to use EnableUnsafeBinaryFormatterSerialization in their aspnet project file.<br>**Impact**: This feature introduces an opt-in method to enable DataContract remoting exception serialization.<br>**Solution**: Remoting V2 introduces DataContract remoting exception serialization. Users can opt this feature by using abricTransportRemotingListenerSettings.ExceptionSerializationTechnique on the Service side and FabricTransportRemotingSettings.ExceptionDeserializationTechnique on the client side.<br>**Documentation Reference**: Read more about it at -<ul><li>[Enabling Data Contract serialization for Remoting exceptions in Service Fabric](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-exception-serialization)</li></ul> |
```

> | 6. | **Feature** | Support creation of End-to-End Developer Experience for Linux development on Windows using WSL2 | **Brief Description**: Provide an updated Service Fabric SDK that will support Linux Local Cluster Management in Windows based on Windows Subsystem for Linux 2 (WSL2) based Ubuntu Virtual Machines (VM).<br>**Impact**: This is an additional feature and will not have any impact on existing behavior.<br>**Solution**: This feature provides a functionality to setup and manage Linux Local Cluster Management in Windows based on WSL2 based Ubuntu VM.<br>**Documentation Reference**: Read more about it at -<ul><li>[Set up Azure Service Fabric Linux cluster on WSL2 linux distribution inside Windows](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-local-linux-cluster-windows-wsl2)</li><li>[Script to setup Service Fabric local cluster for Linux VM using WSL2](https://github.com/microsoft/service-fabric-scripts-and-templates/blob/master/scripts/SetupServiceFabric/SetupServiceFabric.sh)</li></ul> |

Reviewer: ```suggestion
| 6. | **Feature** | Support creation of End-to-End Developer Experience for Linux development on Windows using WSL2 | **Brief Description**: Provide an updated Service Fabric SDK that will support Linux Local Cluster Management in Windows based on Windows Subsystem for Linux 2 (WSL2) based Ubuntu Virtual Machines (VM).<br>**Impact**: This is an additional feature and will not have any impact on existing behavior.<br>**Solution**: This feature provides a functionality to setup and manage Linux Local Cluster Management in Windows based on WSL2 based Ubuntu VMs.<br>**Documentation Reference**: Read more about it at -<ul><li>[Set up Azure Service Fabric Linux cluster on WSL2 linux distribution inside Windows](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-local-linux-cluster-windows-wsl2)</li><li>[Script to setup Service Fabric local cluster for Linux VM using WSL2](https://github.com/microsoft/service-fabric-scripts-and-templates/blob/master/scripts/SetupServiceFabric/SetupServiceFabric.sh)</li></ul> |
```

Reviewer: ^"WSL2 based Ubuntu VM"-->"WSL2 based Ubuntu VMs"

> | 8. | **Feature** | Support for Managed KeyVaultReference | **Brief Description**: Service Fabric's auto-rotation story for application secrets is here: Managed KeyVaultReferences. Leverage Managed Identities for SF Apps, and KeyVaultReferences to get automatic application parameter upgrades as your secrets rotate in the vault.<br>**Documentation Reference**: Read more about it at - <ul><li>[Use KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-keyvaultreferences-in-your-application), and </li><li>[Use Managed KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-managed-keyvaultreferences-in-your-application)</li></ul> |

Reviewer 2: Service Fabric's auto-rotation story for application secrets is here: Managed KeyVaultReferences. -> 

Managed KeyVaultReference is Service Fabric's auto-rotation application secrets.

> | 8. | **Feature** | Support for Managed KeyVaultReference | **Brief Description**: Service Fabric's auto-rotation story for application secrets is here: Managed KeyVaultReferences. Leverage Managed Identities for SF Apps, and KeyVaultReferences to get automatic application parameter upgrades as your secrets rotate in the vault.<br>**Documentation Reference**: Read more about it at - <ul><li>[Use KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-keyvaultreferences-in-your-application), and </li><li>[Use Managed KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-managed-keyvaultreferences-in-your-application)</li></ul> |

Reviewer: ```suggestion
| 8. | **Feature** | Support for Managed KeyVaultReference | **Brief Description**: Service Fabric's auto-rotation story for application secrets is here: Managed KeyVaultReferences. Leverage Managed Identities for SF Apps and KeyVaultReferences to get automatic application parameter upgrades as your secrets rotate in the vault.<br>**Documentation Reference**: Read more about it at - <ul><li>[Use KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-keyvaultreferences-in-your-application), and </li><li>[Use Managed KeyVaultReferences in your application](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-keyvault-references#use-managed-keyvaultreferences-in-your-application)</li></ul> |
```

Reviewer: ^just removed unnecessary comma after "Leverage Managed Identities for SF Apps"

> | 9. | **Feature** | Expose Container ID for currently deployed code packages | **Brief Description**: For processes, Process ID is available from CodePackageEntryPoint structure in both Service Fabric Explorer (SFX) and through Get-ServiceFabricDeployedCodePackage in Powershell. However, for containers, the Container ID is not available via the same mechanism. We can still get this information but it requires multiple docker/containerD API calls to figure out which service fabric service/code package a particular container 'belongs' to. Service Fabric now exposes Container ID in the same fashion that Process ID is exposed.<br>**Solution**: Container ID is added to the CodePakageEntryPoint structure, exposing it in both Service Fabric Explorer (SFX) and through Get-ServiceFabricDeployedCodePackage in Powershell.<br>**Workaround**: Multiple queries to docker/containerD API to get the container ID and decide which is it's corresonding code package. |

Reviewer 2: However, for containers, the Container ID is not available via the same mechanism. We can still get this information but it requires multiple docker/containerD API calls to figure out which service fabric service/code package a particular container 'belongs' to. Service Fabric now exposes Container ID in the same fashion that Process ID is exposed.

-> However, for containers the Container ID was not previously available. Service Fabric will now expose Container ID in the same fashion that Process ID is exposed.

| 10. | **Feature** | Added Fabric_InstanceId environment variable for stateless guest applications | **Brief Description**: Adding Fabric_InstanceId environment variable for stateless guest applications.<br>**Impact**: In order to report health for guest applications replicas using HTTP-based Fabric client from within the guest application, the following information is required PartitionId, ReplicaId (for stateful replicas) and InstanceId (for stateless replicas).<br>Today we already provide PartitionId and ReplicaId (for stateful replicas) as environment variables. Adding the missing InstanceId environment variable for stateless replicas will complete the required information.<br>**Documentation Reference**: To view the list of available Service Fabric environment variable - <ul><li>[Service Fabric environment variables](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-environment-variables-reference)</li></ul> |

> | 11. | **Feature** | Introduced a property in Service Fabric runtime that can be set via SFRP as the ARM resource id | **Brief Description**: To better support Service Fabric Resource Provider (SFRP) & Service Fabric managed clusters (SFMC) scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources.<br>**Impact**: Metadata added for application type are application type provision timestamp and arm resource id. Metadata added for application, service, and cluster are ARM resource id.<br>Existing API's have been updated to read this metadata and support for Powershell, REST, and FabricClient have been added.<br>The runtime created a new API for Service Fabric Resource Provider (SFRP) for updating the metadata. Currently no public support is provided for Powershell since this is only available only for Service Fabric Resource Provider (SFRP) & Service Fabric managed clusters (SFMC) use<br>**Documentation Reference**: Read more about it at - <ul><li>ApplicationTypeMetadata</li><li>ApplicationMetadata</li><li>ServiceMetadata</li></ul> |

Reviewer 2: ID should be capitalized in resource id. 

How does this metadata change impact customers? Can they specify anything new using this API?

> | 11. | **Feature** | Introduced a property in Service Fabric runtime that can be set via SFRP as the ARM resource id | **Brief Description**: To better support Service Fabric Resource Provider (SFRP) & Service Fabric managed clusters (SFMC) scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources.<br>**Impact**: Metadata added for application type are application type provision timestamp and arm resource id. Metadata added for application, service, and cluster are ARM resource id.<br>Existing API's have been updated to read this metadata and support for Powershell, REST, and FabricClient have been added.<br>The runtime created a new API for Service Fabric Resource Provider (SFRP) for updating the metadata. Currently no public support is provided for Powershell since this is only available only for Service Fabric Resource Provider (SFRP) & Service Fabric managed clusters (SFMC) use<br>**Documentation Reference**: Read more about it at - <ul><li>ApplicationTypeMetadata</li><li>ApplicationMetadata</li><li>ServiceMetadata</li></ul> |

Reviewer: ```suggestion
| 11. | **Feature** | Introduced a property in Service Fabric runtime that can be set via SFRP as the ARM resource id | **Brief Description**: To better support Service Fabric Resource Provider (SFRP) and Service Fabric managed clusters (SFMC) scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources.<br>**Impact**: Metadata added for application type are application type provision timestamp and arm resource id. Metadata added for application, service, and cluster are ARM resource id.<br>Existing API's have been updated to read this metadata and support for Powershell, REST, and FabricClient have been added.<br>The runtime created a new API for Service Fabric Resource Provider (SFRP) for updating the metadata. Currently, no public support is provided for Powershell since this is only available only for Service Fabric Resource Provider (SFRP) & Service Fabric managed clusters (SFMC) use<br>**Documentation Reference**: Read more about it at - <ul><li>ApplicationTypeMetadata</li><li>ApplicationMetadata</li><li>ServiceMetadata</li></ul> |
```

Reviewer: ^"&"-->"and" and comma after Currently

> | 12. | **Feature** | Exposed application type provision timestamp | **Brief Description**: Added application type provision timestamp that is exposed via existing application type API's.<br>**Impact**: To better support Service Fabric Resource Provider (SFRP) and Service Fabric managed clusters (SFMC) scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources.<br>**Documentation Reference**: Read more about it at - <ul><li>ApplicationTypeMetadata</li></ul> |

Reviewer 2: Can customers fetch this timestamp to determine when their application deployments completed?

> | 13. | **Feature** | Exposed API for reporting MoveCost | **Brief Description**: Exposed API for reporting MoveCost on behalf of other service replicas and partitions.<br>**Impact**: MoveCost could be changed with action outside of partition. If clients don't own service logic, they could use only static load.<br>**Solution**: New API allows reporting MoveCost from outside of the service, which could be beneficial for clients that don't have service logic.<br>**Documentation Reference**: Read more about it at - <ul><li>[Reporting move cost for a partition](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-resource-manager-movement-cost#reporting-move-cost-for-a-partition)</li></ul>**Workaround**: Replicas could report MoveCost values for themselves or clients could specify static MoveCost value during service creation. |

Reviewer 2: What can customers do with this move cost value? Can they prevent certain replicas and partitions from being moved within the cluster?

> | 14. | **Feature** | Create replacement replicas before Replica Restart Wait Duration (RRWD) expires when there are less than min replicas up | **Brief Description**: Starting with Service Fabric version 9.0, we have ability to adjust logic around Replica Restart Wait Duration.<br>Currently, when replica goes down, SF waits for a configurable time period until decides to build a new replica elsewhere, hoping the existing replica might become available again.<br>The problem here is that, when partition goes below MinReplicaSetSize number of replicas, it will still wait for that period of time before bringing new replica up, causing unavailability for the customer.<br>By toggling IgnoreReplicaRestartWaitDurationWhenBelowMinReplicaSetSize on, SF will not wait for problematic replica, but start building new one instantly, making sure that there is at least MinReplicaSetSize replicas available as soon as possible.<br>This configuration is dynamic, and customer can toggle the behavior accordingly.<br>The default value for this config is false, meaning that it is switched off.<br>**Impact**: SQL MI asked for the feature as a way to increase availability.<br> |

Reviewer 2: SQL MI asked for the feature as a way to increase availability. -> Is this something that we would want to expose externally?

> | 15. | **Feature** | Enforce a configurable Max value on the InstanceCloseDelayDuration | **Brief Description**: Enforcing a configurable Max value on the InstanceCloseDelayDuration to prevent users from inadvertently specifying huge, delay values which could result in updates getting stuck.<br>Checks have been added in service creation/updation, app upgrade, application & fabric upgrade overrides to prevent users from configuring anything more than the max config value.<br>The max value is a cluster configuration, defaulting to 30mins, so an InstanceCloseDelay can't be specified more than this value. Users can modify this value to anything as they require in the cluster config.<br>For already misconfigured services, the delay would be ceiled to the max config value.<br>**Documentation Reference**: For more about it at <ul><li>[Service Fabric application upgrade: Advanced topics](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-upgrade-advanced)</li></ul>**Breaking Change**: This would be a **breaking change** if customers are already configuring more than 30 mins delay.<br>**Workaround**: They would need to define a higher value in the cluster config. |

Reviewer: ```suggestion
| 15. | **Feature** | Enforce a configurable Max value on the InstanceCloseDelayDuration | **Brief Description**: Enforcing a configurable Max value on the InstanceCloseDelayDuration to prevent users from inadvertently specifying huge delay values which could result in updates getting stuck.<br>Checks have been added in service creation/updation, app upgrade, application and fabric upgrade overrides to prevent users from configuring anything more than the max config value.<br>The max value is a cluster configuration, defaulting to 30mins, so an InstanceCloseDelay can't be specified more than this value. Users can modify this value to anything they require in the cluster config.<br>For already misconfigured services, the delay would be ceiled to the max config value.<br>**Documentation Reference**: For more about it at <ul><li>[Service Fabric application upgrade: Advanced topics](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-upgrade-advanced)</li></ul>**Breaking Change**: This would be a **breaking change** if customers are already configuring more than 30 mins delay.<br>**Workaround**: They would need to define a higher value in the cluster config. |
```

Reviewer: ^rephrasing: "&"-->"and", "anything as they require"-->"anything they require", removed unnecessary comma after "huge"

> | 16. | **Feature** | Added ability to enumerate actor reminders | **Brief Description**: Exposed an API to list all the reminders in the actor service.GetReminderAsync() API has been added to IActorService to enable users to enumerate reminders. <br>**Solution**: Actor customers now has the ability to enumerate the reminders in their service. If a null value is passed to the actorId filter, then the Api returns list of all the reminders. If a valid actor id is passed, then all the reminders corresponding to the actor id input are returned.<table><tr><td><pre lang="csharp">Task<ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>> GetRemindersAsync(<br>    ActorId actorId,<br>    ContinuationToken continuationToken,<br>    CancellationToken cancellationToken);</pre></td></tr><tr><td></table>The page size of the result can be controlled by setting the PageSize static field of the ReminderPagedResult inside the actor service.<table><tr><td><pre lang="csharp">ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>.SetDefaultPageSize(/*new size*/)</pre></td></tr></table>**Documentation Reference**: Read more about it at -<ul><li>[IActorService.GetReminderAsync(ActorId, ContinuationToken, CancellationToken) Method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicefabric.actors.iactorservice.getreminderasync)</li></ul> |

Reviewer 2: Consistency in capitalization -> API

> | 16. | **Feature** | Added ability to enumerate actor reminders | **Brief Description**: Exposed an API to list all the reminders in the actor service.GetReminderAsync() API has been added to IActorService to enable users to enumerate reminders. <br>**Solution**: Actor customers now has the ability to enumerate the reminders in their service. If a null value is passed to the actorId filter, then the Api returns list of all the reminders. If a valid actor id is passed, then all the reminders corresponding to the actor id input are returned.<table><tr><td><pre lang="csharp">Task<ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>> GetRemindersAsync(<br>    ActorId actorId,<br>    ContinuationToken continuationToken,<br>    CancellationToken cancellationToken);</pre></td></tr><tr><td></table>The page size of the result can be controlled by setting the PageSize static field of the ReminderPagedResult inside the actor service.<table><tr><td><pre lang="csharp">ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>.SetDefaultPageSize(/*new size*/)</pre></td></tr></table>**Documentation Reference**: Read more about it at -<ul><li>[IActorService.GetReminderAsync(ActorId, ContinuationToken, CancellationToken) Method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicefabric.actors.iactorservice.getreminderasync)</li></ul> |

Reviewer: ```suggestion
| 16. | **Feature** | Added ability to enumerate actor reminders | **Brief Description**: Exposed an API to list all the reminders in the actor service.GetReminderAsync() API has been added to IActorService to enable users to enumerate reminders. <br>**Solution**: Actor customers now have the ability to enumerate the reminders in their service. If a null value is passed to the actorId filter, then the API returns list of all the reminders. If a valid actor ID is passed, then all the reminders corresponding to the actor ID input are returned.<table><tr><td><pre lang="csharp">Task<ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>> GetRemindersAsync(<br>    ActorId actorId,<br>    ContinuationToken continuationToken,<br>    CancellationToken cancellationToken);</pre></td></tr><tr><td></table>The page size of the result can be controlled by setting the PageSize static field of the ReminderPagedResult inside the actor service.<table><tr><td><pre lang="csharp">ReminderPagedResult<KeyValuePair<ActorId, List<ActorReminderState>>>.SetDefaultPageSize(/*new size*/)</pre></td></tr></table>**Documentation Reference**: Read more about it at -<ul><li>[IActorService.GetReminderAsync(ActorId, ContinuationToken, CancellationToken) Method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicefabric.actors.iactorservice.getreminderasync)</li></ul> |
```

Reviewer: ^"Actor customers now has"-->"Actor customers now have", capitalized second instance of API, capitalized two instances of ID

> | 17. | **Feature** | Made updates to MR UI | **Brief Description**: This update adds new page under each Infrastructure Service section which displays information related to undergoing Infrastructure updates on the Virtual Machine Scale Sets. Users can get information about currently Executing updates, Pending updates and the State of each update right from the Service fabric Explorer.<br>**Impact**: This helps in quickly identifying the undergoing infrastructure updates and understanding overall health of the Cluster.<br>Correlating Infrastructure updates with corresponding repair Tasks , to identify why the updates are stuck.<br>**Documentation Reference**: Read more about it at -<ul><li>[Introduction to the Service Fabric Infrastructure Service](https://docs.microsoft.com/en-us/azure/service-fabric/infrastructure-service-faq)</li><li>[Azure Service Fabric node types and virtual machine scale sets](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-nodetypes)</li></ul> |

Reviewer 2: We don't expose MR externally -> I think this needs to be referred to as platform events.

> | 17. | **Feature** | Made updates to MR UI | **Brief Description**: This update adds new page under each Infrastructure Service section which displays information related to undergoing Infrastructure updates on the Virtual Machine Scale Sets. Users can get information about currently Executing updates, Pending updates and the State of each update right from the Service fabric Explorer.<br>**Impact**: This helps in quickly identifying the undergoing infrastructure updates and understanding overall health of the Cluster.<br>Correlating Infrastructure updates with corresponding repair Tasks , to identify why the updates are stuck.<br>**Documentation Reference**: Read more about it at -<ul><li>[Introduction to the Service Fabric Infrastructure Service](https://docs.microsoft.com/en-us/azure/service-fabric/infrastructure-service-faq)</li><li>[Azure Service Fabric node types and virtual machine scale sets](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-nodetypes)</li></ul> |

Reviewer: ```suggestion
| 17. | **Feature** | Made updates to MR UI | **Brief Description**: This update adds new page under each Infrastructure Service section which displays information related to undergoing Infrastructure updates on the Virtual Machine Scale Sets. Users can get information about currently Executing updates, Pending updates and the State of each update right from the Service Fabric Explorer.<br>**Impact**: This helps in quickly identifying the undergoing infrastructure updates and understanding overall health of the Cluster.<br>Correlating Infrastructure updates with corresponding repair Tasks to identify why the updates are stuck.<br>**Documentation Reference**: Read more about it at -<ul><li>[Introduction to the Service Fabric Infrastructure Service](https://docs.microsoft.com/en-us/azure/service-fabric/infrastructure-service-faq)</li><li>[Azure Service Fabric node types and virtual machine scale sets](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-nodetypes)</li></ul> |
```

Reviewer: ^removed unnecessary space and comma after Tasks and capitalized Fabric

> | 18. | **Feature** | Add Central Secret Store(CSS) to SystemServicesForScale | **Brief Description**: Instantiate the system service Central Secret Store (CSS) with good defaults, omitting the need for additional user-provided configuration settings.<br>**Impact**: Users previously had to specify configuration settings upon activating the Central Secret Store service; failing to specify these settings led to cluster upgrade failures. After this change, enabling the service proceeds with safe defaults for these settings.<br>**Solution**: Add Central Secret Store to the SystemServicesForScale sections of WRPSettings.xml in the WindowsFabric repository, and to Settings.xml in the winfab_rp repository.<br>**Documentation Reference**: Read more about it at -<ul><li>[Enable Central Secrets Service](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-secret-store#enable-central-secrets-service)</li></ul>**Breaking Change**: Attempting to manually set the MinReplicaCount and TargetReplicaCount configuration settings of the Central Secret Store service will be overridden with default values as determined by the runtime for any system service.<br>**Workaround**: [WORKAROUND] |

Reviewer 2: What does good defaults mean? Can it be expanded on what we are setting by default?

> | 18. | **Feature** | Add Central Secret Store(CSS) to SystemServicesForScale | **Brief Description**: Instantiate the system service Central Secret Store (CSS) with good defaults, omitting the need for additional user-provided configuration settings.<br>**Impact**: Users previously had to specify configuration settings upon activating the Central Secret Store service; failing to specify these settings led to cluster upgrade failures. After this change, enabling the service proceeds with safe defaults for these settings.<br>**Solution**: Add Central Secret Store to the SystemServicesForScale sections of WRPSettings.xml in the WindowsFabric repository, and to Settings.xml in the winfab_rp repository.<br>**Documentation Reference**: Read more about it at -<ul><li>[Enable Central Secrets Service](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-secret-store#enable-central-secrets-service)</li></ul>**Breaking Change**: Attempting to manually set the MinReplicaCount and TargetReplicaCount configuration settings of the Central Secret Store service will be overridden with default values as determined by the runtime for any system service.<br>**Workaround**: [WORKAROUND] |

Reviewer: ```suggestion
| 18. | **Feature** | Add Central Secret Store (CSS) to SystemServicesForScale | **Brief Description**: Instantiate the system service CSS with good defaults, omitting the need for additional user-provided configuration settings.<br>**Impact**: Users previously had to specify configuration settings upon activating the CSS service; failing to specify these settings led to cluster upgrade failures. After this change, enabling the service proceeds with safe defaults for these settings.<br>**Solution**: Add CSS to the SystemServicesForScale sections of WRPSettings.xml in the WindowsFabric repository and to Settings.xml in the winfab_rp repository.<br>**Documentation Reference**: Read more about it at -<ul><li>[Enable Central Secrets Service](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-secret-store#enable-central-secrets-service)</li></ul>**Breaking Change**: Attempting to manually set the MinReplicaCount and TargetReplicaCount configuration settings of the CSS service will be overridden with default values as determined by the runtime for any system service.<br>**Workaround**: [WORKAROUND] |
```

> | 19. | **Feature** | Support adding Service Fabric Resource Provider (SFRP) metadata to app type + version entities, starting with ARM resource Id | **Brief Description**: Added a new API that can be used to add SFRP metadata to app type entities.<br>**Impact**: To better support SFRP & SFMC scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources. |

Reviewer 2: I would recommend grouping all of these SFRP metadata changes side by side.

> | 19. | **Feature** | Support adding Service Fabric Resource Provider (SFRP) metadata to app type + version entities, starting with ARM resource Id | **Brief Description**: Added a new API that can be used to add SFRP metadata to app type entities.<br>**Impact**: To better support SFRP & SFMC scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources. |

Reviewer: ```suggestion
| 19. | **Feature** | Support added for Service Fabric Resource Provider (SFRP) metadata to app type + version entities, starting with ARM resource Id | **Brief Description**: Added a new API that can be used to add SFRP metadata to app type entities.<br>**Impact**: To better support SFRP and SFMC scenarios, SF runtime needs to allow support for additional metadata for ARM deployed resources. |
```

Reviewer: ^rephrasing "Support adding Service Fabric"-->"Support added for Service..."; "&"-->"and"

> | 20. | **Bug Fix** | Managed Identity Token Service (MITS) fails renegotiation with gateway on systems with TLS 1.3 enabled | **Brief Description**: The Managed Identity Token Service (MITS) will lose its connection to the gateway when deployed on Win2022, and won't attempt to renegotiate it.<br>**Impact**: On Windows 2022 or other situations where TLS1.3 is available/the default, MITS will fail to renegotiate its inbound connections, and will fail all inbound requests. SF-hosted user services attempting to acquire tokens representing their managed identities will experience timeouts.<br>**Fix**: Limit MITS to using TLS 1.2 only, until Windows 2022 becomes a supported OS for Service Fabric.<br>**Workaround**: Note Service Fabric does not list Windows 2022 as a supported OS version at this time - primarily because it defaults to TLS1.3, and the SF runtime needs additional fixes to support this scenario. Our recommendation is to stick with the officially supported OS images/versions. In this particular case, restarting MITS on the failing node will work momentarily, which may be sufficient for an application to obtain its tokens. However, the failing behavior will re-occur within the order of minutes. |

Reviewer: ```suggestion
| 20. | **Bug Fix** | Managed Identity Token Service (MITS) fails renegotiation with gateway on systems with TLS 1.3 enabled | **Brief Description**: MITS will lose its connection to the gateway when deployed on Win2022 and won't attempt to renegotiate it.<br>**Impact**: On Windows 2022 or other situations where TLS1.3 is available/the default, MITS will fail to renegotiate its inbound connections and will fail all inbound requests. SF-hosted user services attempting to acquire tokens representing their managed identities will experience timeouts.<br>**Fix**: Limit MITS to using TLS 1.2 only, until Windows 2022 becomes a supported OS for Service Fabric.<br>**Workaround**: Note Service Fabric does not list Windows 2022 as a supported OS version at this time - primarily because it defaults to TLS1.3, and the SF runtime needs additional fixes to support this scenario. Our recommendation is to stick with the officially supported OS images/versions. In this particular case, restarting MITS on the failing node will work momentarily, which may be sufficient for an application to obtain its tokens. However, the failing behavior will reoccur within the order of minutes.
```

Reviewer: ^replaced second instance of Managed Identity Token Service with MITS; removed unnecessary commas; removed unnecessary hyphen

> | 21. | **Bug Fix** | Validation fails when RunToCompletion is specified and ServicePackageActivationMode is SharedProcess | **Brief Description**: New services, including both default services for new applications and new services for existing applications, will fail validation when they are set to RunToCompletion and ServicePackageActivationMode is SharedProcess. Such cases have never been supported by Service Fabric, however, we would accept this services and not honor RunToCompletion, so they would restart on exit.<br>**Impact**: Before, we would accept services set to RunToCompletion where ServicePackageActivationMode was SharedProcess, yet we did not honor RunToCompletion. Now validation will fail in that scenario and all services set to RunToCompletion will honor that execution policy.<br>**Fix**: For default services, validation fails during ApplicationManifest validation. For new services on existing applications, the cluster manager (CM) fails the new service validation before creating it.<br>**Documentation Reference**: Read more about it at -<ul><li>[RunToCompletion](https://docs.microsoft.com/en-us/azure/service-fabric/run-to-completion)</li></ul>**Breaking Change**: New applications and application upgrades where there are default services that RunToCompletion and ServicePackageActivationMode is SharedProcess will fail. New services for existing applications will also fail for such cases. |

Reviewer: ```suggestion
| 21. | **Bug Fix** | Validation fails when RunToCompletion is specified and ServicePackageActivationMode is SharedProcess | **Brief Description**: New services, including both default services for new applications and new services for existing applications, will fail validation when they are set to RunToCompletion and ServicePackageActivationMode is SharedProcess. Such cases have never been supported by Service Fabric; however, we would accept these services and not honor RunToCompletion, so they would restart on exit.<br>**Impact**: Before, we would accept services set to RunToCompletion where ServicePackageActivationMode was SharedProcess, yet we did not honor RunToCompletion. Now validation will fail in that scenario and all services set to RunToCompletion will honor that execution policy.<br>**Fix**: For default services, validation fails during ApplicationManifest validation. For new services on existing applications, the cluster manager (CM) fails the new service validation before creating it.<br>**Documentation Reference**: Read more about it at -<ul><li>[RunToCompletion](https://docs.microsoft.com/en-us/azure/service-fabric/run-to-completion)</li></ul>**Breaking Change**: New applications and application upgrades where there are default services that RunToCompletion and ServicePackageActivationMode is SharedProcess will fail. New services for existing applications will also fail for such cases. |
```

Reviewer: ^"this services"-->"these services"; replaced comma with semicolon before "however"

| 22. | **Bug Fix** | IActorService.GetActors fails in KVS due to case sensitivity | **Brief Description**: GetActors API returns a single page infinitely if a KVS actor service has upper case actors as well as lower case actors.<br>**Impact**: GetActors enumeration API<br>**Fix**: GetActors API will enumerate all actors in the KVS service with upper and lower case actors |

<br>

## Future Retirement and Deprecation Path Callouts

| Support | Option | Details |
|-----------------|--------|---------|

> | **Retired support for ASP.NET Core 1.0** | Migrate usage to ASP.NET Core 2.1 or greater | Service Fabric ASP.NET Core packages are built against ASP.NET Core 1.0 binaries which are out of support. Starting Service Fabric 8.2, we will be building Service Fabric ASP.NET Core packages against .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, .NET Framework 4.6.1. As a result, they can be used only to build services targeting .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, >=.NET Framework 4.6.1 respectively.<br>For .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, Service Fabric ASP.NET Core will be taking dependency on Microsoft.AspNetCore.App shared framework, whereas for NetFx target frameworks >=.NET Framework 4.6.1, Service Fabric ASP.NET Core will be taking dependency on ASP.NET Core 2.1 packages.<br>The package Microsoft.ServiceFabric.AspNetCore.WebListener will no longer be shipped, Microsoft.ServiceFabric.AspNetCore.HttpSys package should be used instead.|

Reviewer: ```suggestion
| **Retired support for ASP.NET Core 1.0** | Migrate usage to ASP.NET Core 2.1 or greater | Service Fabric ASP.NET Core packages are built against ASP.NET Core 1.0 binaries, which are out of support. Starting Service Fabric 8.2, we will be building Service Fabric ASP.NET Core packages against .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, .NET Framework 4.6.1. As a result, they can be used only to build services targeting .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, >=.NET Framework 4.6.1 respectively.<br>For .NET Core 3.1, .NET Core 5.0, .NET Core 6.0, Service Fabric ASP.NET Core will be taking dependency on Microsoft.AspNetCore.App shared framework; whereas, for NetFx target frameworks >=.NET Framework 4.6.1, Service Fabric ASP.NET Core will be taking dependency on ASP.NET Core 2.1 packages.<br>The package Microsoft.ServiceFabric.AspNetCore.WebListener will no longer be shipped, Microsoft.ServiceFabric.AspNetCore.HttpSys package should be used instead.|
```

Reviewer: ^added two commas; replaced one comma with semicolon

> | **Deprecated support for .NET Core 2.x** | Migrate usage to .NET Core 3.1 | .NET Core runtime LTS 2.1 runtime is out of support as of Aug 21, 2021, and .NET Core runtime 2.2 is out of support from Dec 2019.<br>Service Fabric releases after those dates will drop support for Service Fabric applications running with .NET Core 2.1 & .NET Core 2.2 runtimes respectively.<br>Current applications running on .NET Core 2.x runtime will continue to work, but requests for investigations or request for changes will no longer receive support. Service Fabric .NET SDK will take a dependency on .NET runtime 3.x features to support Service Fabric .NET Core applications. |

Reviewer: ```suggestion
| **Deprecated support for .NET Core 2.x** | Migrate usage to .NET Core 3.1 | .NET Core runtime LTS 2.1 runtime is out of support as of Aug 21, 2021, and .NET Core runtime 2.2 is out of support since Dec 2019.<br>Service Fabric releases after those dates will drop support for Service Fabric applications running with .NET Core 2.1 & .NET Core 2.2 runtimes respectively.<br>Current applications running on .NET Core 2.x runtime will continue to work, but requests for investigations or request for changes will no longer receive support. Service Fabric .NET SDK will take a dependency on .NET runtime 3.x features to support Service Fabric .NET Core applications. |
```

Reviewer: ^"from"-->"since"

> | **(Upcoming) Deprecation of support for .NET 5.0** | Migrate usage to .NET 6.0 or .NET Core 3.1 |.NET 5.0 runtime is reaching end of life on May 8, 2022.<br>Service Fabric releases after that date will drop support for Service Fabric applications running with .NET 5.0 runtime.<br>Current applications running on .NET 5.0 runtime will continue to work, but requests for investigations or request for changes will no longer receive support. |

Reviewer: ```suggestion
| **(Upcoming) Deprecation of support for .NET 5.0** | Migrate usage to .NET 6.0 or .NET Core 3.1 |.NET 5.0 runtime is reaching end-of-life on May 8, 2022.<br>Service Fabric releases after that date will drop support for Service Fabric applications running with .NET 5.0 runtime.<br>Current applications running on .NET 5.0 runtime will continue to work, but requests for investigations or request for changes will no longer receive support. |
```

Reviewer: ^"end of life"-->"end-of-life"

> | **(Upcoming) Retirement of support for Ubuntu 16.04** | Upgrade to more recent 18.04 LTS or 20.04 LTTS | Ubuntu 16.04 LTS reached it's 5-year end-of-life window on April 30, 2021.<br>Service Fabric runtime has dropped support for 16.04 LTS and we recommend moving your clusters and applications to Ubuntu 18.04 LTS or 20.04 LTS.<br>Current applications running on it will continue to work, but requests for investigations or requests for change will no longer receive support.<br>Service Fabric runtime will soon be permenantly retiring support for Ubuntu 16.04 LTS and will no longer make packages available for it. |

Reviewer: ```suggestion
| **(Upcoming) Retirement of support for Ubuntu 16.04** | Upgrade to more recent 18.04 LTS or 20.04 LTTS | Ubuntu 16.04 LTS reached it's 5-year end-of-life window on April 30, 2021.<br>Service Fabric runtime has dropped support for 16.04 LTS, and we recommend moving your clusters and applications to Ubuntu 18.04 LTS or 20.04 LTS.<br>Current applications running on it will continue to work, but requests for investigations or requests for change will no longer receive support.<br>Service Fabric runtime will soon be permenantly retiring support for Ubuntu 16.04 LTS and will no longer make packages available for it. |
```

Reviewer: ^added comma

> | **(Upcoming) Deprecation of support for BinaryFormatter based remoting exception serialization** | Migrate usage to Data Contract Serilization (DCS) based remoting exception. | .NET Framework team strongly recommends against using BinaryFormatter class due to security vulnerabilities and has marked it's serialization & deserielization methods as obsolete.<br>Current applications using it will continue to work as-is, but Service Fabric strongly recommends moving to using Data Contract Serilization (DCS) based remoting exception instead.<br>Service Fabric runtime will soon be stop using BinaryFormatter based remoting exception serialization by default and move to using Data Contract Serilization (DCS) based remoting exception serialization by default. |

Reviewer: ```suggestion
| **(Upcoming) Deprecation of support for BinaryFormatter based remoting exception serialization** | Migrate usage to Data Contract Serialization (DCS) based remoting exception. | .NET Framework team strongly recommends against using BinaryFormatter class due to security vulnerabilities and has marked its serialization and deserialization methods as obsolete.<br>Current applications using it will continue to work as-is, but Service Fabric strongly recommends moving to using DCS based remoting exception instead.<br>Service Fabric runtime will soon be stop using BinaryFormatter based remoting exception serialization by default and move to using DCS based remoting exception serialization by default. |
```

Reviewer: ^later Data Contract Serialization-->DCS; fixed serialization/deserialization typos; it's --> its; & --> and

> | **(Upcoming) Archival & the removal of older version of Service Fabric packages from the Download Center** | Migrate to more recent and supported versions of Service Fabric runtime | Service Fabric runtime will soon be archiving and removing Service Fabric runtime version 6.4 packages and older, as well as SDK version 3.3 packages and older from the package Down Center. |

Reviewer: & --> and; removed unnecessary the

<br>

## Current Breaking Changes
* **Guest Executable and Container Applications**: Guest executable and container applications created or upgraded in Service Fabric clusters with runtime versions 7.1+ are incompatible with prior Service Fabric runtime versions (e.g. Service Fabric 7.0).<br/>
    Following scenarios are impacted:<br/>
    * An application with guest executables or containers is created or upgraded in an Service Fabric 7.1+ cluster.<br/>
    The cluster is then downgraded to a previous Service Fabric runtime version (e.g. Service Fabric 7.0).<br/>
    The application fails to activate.<br/>
    * A cluster upgrade from pre-SF 7.1 version to Service Fabric 7.1+ version is in progress.<br/>
    In parallel with the Service Fabric runtime upgrade, an application with guest executables or containers is created or upgraded.<br/>
    The Service Fabric runtime upgrade starts rolling back (due to any reason) to the pre-SF 7.1 version.<br/>
    The application fails to activate.<br/>

>     To avoid issues when upgrading from a pre-SF 7.1 runtime version to an Service Fabric 7.1+ runtime version, do not create or upgrade applications with guest executables or containers while the Service Fabric runtime upgrade is in progress.<br/>

Reviewer: ```suggestion
    To avoid issues when upgrading from a pre-SF 7.1 runtime version to a Service Fabric 7.1+ runtime version, do not create or upgrade applications with guest executables or containers while the Service Fabric runtime upgrade is in progress.<br/>
```

Reviewer: ^an Service Fabric --> a Service Fabric

<br>

## Service Fabric Add-ons Applications

**FabricHealer - Auto-Mitigation in Service Fabric Clusters**

[FabricHealer](https://github.com/microsoft/service-fabric-healer) is a Service Fabric application that attempts to automatically fix a set of reliably solvable problems that can take place in Service Fabric application services, including containers, host virtual machines, and logical disks (scoped to space usage problems only). These repairs mostly employ a set of Service Fabric API calls, but can also be fully customizable (like Disk repair). All repairs are safely orchestrated through Service Fabric’s RepairManager service. Repair configuration is written as Prolog-like logic with [supporting external predicates](https://github.com/microsoft/service-fabric-healer/tree/main/FabricHealer/Repair/Guan) written in C#. This is made possible by a new logic programming library for .NET, [Guan](https://github.com/microsoft/guan). The fun starts when FabricHealer detects supported error or warning states reported by [FabricObserver](https://github.com/microsoft/service-fabric-observer). FabricHealer requires FabricObserver to be running in the same cluster.

> We are especially interested in Feedback with respect to the new [Configuration-as-Logic](https://github.com/microsoft/service-fabric-healer/blob/main/Documentation/LogicWorkflows.md) feature. Out of the box, FabricHealer is configured to solve several problems automatically. You do not have to be a logic programmer to use FabricHealer right away. Check out the [scenario-based getting started samples](https://github.com/microsoft/service-fabric-healer/blob/main/Documentation/Using.md) to get going quickly. There are several logic rules supplied by default that span service to VM level auto-mitigation. You just need to change arguments in the rules to match your needs. 

Reviewer: ```suggestion
We are especially interested in feedback with respect to the new [Configuration-as-Logic](https://github.com/microsoft/service-fabric-healer/blob/main/Documentation/LogicWorkflows.md) feature. Out of the box, FabricHealer is configured to solve several problems automatically. You do not have to be a logic programmer to use FabricHealer right away. Check out the [scenario-based getting started samples](https://github.com/microsoft/service-fabric-healer/blob/main/Documentation/Using.md) to get going quickly. There are several logic rules supplied by default that span service to VM level auto-mitigation. You just need to change arguments in the rules to match your needs. 
```

Reviewer: ^lower case feedback
